builtin/rebase.c

```diff
@@ -0,0 +1,687 @@
+	}
+
+	tree = parse_tree_indirect(oid);
+	prime_cache_tree(the_repository->index, tree);
+
+	if (write_locked_index(the_repository->index, &lock, COMMIT_LOCK) < 0)
+		ret = error(_("could not write index"));
+	free((void *)desc.buffer);
+
+	if (ret)
+		return ret;
+
+	reflog_action = getenv(GIT_REFLOG_ACTION_ENVIRONMENT);
+	strbuf_addf(&msg, "%s: ", reflog_action ? reflog_action : "rebase");
+	prefix_len = msg.len;
+
+	if (!get_oid("ORIG_HEAD", &oid_old_orig))
+		old_orig = &oid_old_orig;
+	if (!get_oid("HEAD", &oid_orig)) {
+		orig = &oid_orig;
+		strbuf_addstr(&msg, "updating ORIG_HEAD");
+		update_ref(msg.buf, "ORIG_HEAD", orig, old_orig, 0,
+			   UPDATE_REFS_MSG_ON_ERR);
+	} else if (old_orig)
+		delete_ref(NULL, "ORIG_HEAD", old_orig, 0);
+	strbuf_setlen(&msg, prefix_len);
+	strbuf_addstr(&msg, "updating HEAD");
+	ret = update_ref(msg.buf, "HEAD", oid, orig, 0, UPDATE_REFS_MSG_ON_ERR);
+
+	strbuf_release(&msg);
+
+	return ret;
+}
+
+int cmd_rebase(int argc, const char **argv, const char *prefix)
+{
+	enum {
+		NO_ACTION,
+		ACTION_CONTINUE,
+		ACTION_SKIP,
+		ACTION_ABORT,
+		ACTION_QUIT,
+		ACTION_EDIT_TODO,
+		ACTION_SHOW_CURRENT_PATCH
+	} action = NO_ACTION;
+	struct rebase_options options = { REBASE_UNSPECIFIED };
+	const char *branch_name = NULL;
+	int ret, flags, quiet = 0, total_argc = 0, in_progress = 0;
+	int ok_to_skip_pre_rebase = 1;
+	int keep_empty = 1;
+	int autosquash = 0;
+	int fork_point = 0;
+	int do_merge = 1;
+	struct strbuf msg = STRBUF_INIT;
+	struct strbuf revisions = STRBUF_INIT;
+	struct strbuf git_am_opt = STRBUF_INIT;
+	struct strbuf buf = STRBUF_INIT;
+	const char *restrict_revision = NULL;
+	struct object_id merge_base;
+	struct option builtin_rebase_options[] = {
+		OPT_STRING(0, "onto", &options.onto_name,
+			N_("revision"),
+			N_("rebase onto given branch instead of upstream")),
+		OPT_BOOL(0, "no-verify", &ok_to_skip_pre_rebase,
+			N_("allow pre-rebase hook to run")),
+		OPT_NEGBIT('q', "quiet", &options.flags,
+			N_("be quiet. implies --no-stat"),
+			REBASE_VERBOSE | REBASE_DIFFSTAT),
+		OPT_BIT('v', "verbose", &options.flags,
+			N_("display a diffstat of what changed upstream"),
+			REBASE_VERBOSE | REBASE_DIFFSTAT),
+		OPT_CMDMODE(0, "continue", &action, N_("continue"),
+			ACTION_CONTINUE),
+		OPT_CMDMODE(0, "abort", &action,
+			    N_("abort and check out the original branch"),
+			    ACTION_ABORT),
+		OPT_CMDMODE(0, "skip", &action,
+			    N_("skip current patch and continue"), ACTION_SKIP),
+		OPT_CMDMODE(0, "edit-todo", &action, N_("edit the todo list "
+			    "during an interactive rebase"), ACTION_EDIT_TODO),
+		OPT_CMDMODE(0, "quit", &action,
+			    N_("abort but keep HEAD where it is"), ACTION_QUIT),
+		OPT_CMDMODE(0, "show-current-patch", &action,
+			    N_("show the patch file being applied or merged"),
+			    ACTION_SHOW_CURRENT_PATCH),
+		OPT_BOOL(0, "no-keep-empty", &keep_empty,
+			N_("preserve empty commits during rebase")),
+		OPT_BOOL(0, "no-autosquash", &autosquash,
```

Comment:
So now you parse the `--autosquash` option. But why `"no-autosquash"`? That *inverts* the meaning of `autosquash`: it would now be set to `1` (corresponding to the Boolean `true`) when the user specified the `--no-autosquash` option on the command-line option! That is the exact opposite of what we want.

Besides, maybe you should wait with converting this until you have had the time to finish your previous work on hacking up `sequencer.c`'s `apply_autosquash()` function to make it reusable from `builtin/rebase.c`. Because it is a bit pointless to implement option parsing for an option without then also letting the builtin rebase *act* on it.